File: js/linkedList.js
function MyException(message) {
   this.message = message;
   this.name = 'MyException';
}

function Node(value) {
  this.value = value;
  this.next = null;
  this.prev = null;
}

function LinkedList() {
  this.length = 0;
  this.head = null;
  this.tail = null;
}

LinkedList.prototype.addNode = function(node) {
  if (!(node instanceof Node)) throw new MyException('input is not a Node');
  this.length += 1;
  if (!this.head) {
    this.head = node;
    this.tail = node;    
  } else {
    node.prev = this.tail;
    this.tail.next = node;
    this.tail = node;
  }
  return this;
};

LinkedList.prototype.removeLast = function() {
  let node = this.head;
  if(!node) throw new MyException('Linked List is already empty');
  if (!node.next) {
    this.head = null;
    this.tail = null;
  } else {
    while (node.next) {
      node = node.next;
    }
    node.prev.next = null;
    this.tail = node.prev;
  }
  this.length -= 1;
  return this;
};

LinkedList.prototype.clean = function() {
  let node = this.tail;
  if(!node) throw new MyException('Linked List is already empty');
  function recursiveRm(node) {
    if (node.next) node.next.prev = null;
    node.next = null;
    node.value = null;
    this.length -= 1;
    if (!node.prev) {
      node.value = null;
      return 0;
    }
    return recursiveRm.call(this, node.prev);
  }
  recursiveRm.call(this, node);
  this.head = null;
  this.tail = null;
  return this;
};

LinkedList.prototype.getNode = function(index, notSkip=true) {
  let node = this.head;
  if(notSkip && !node) throw new MyException('Linked List is already empty');
  if (index > this.length || index < 0) throw new MyException('invalid index input');
  for (let i = 0; i < index; i += 1) {
    node = node.next;
  }

  return node;
};

LinkedList.prototype.insertAtIndex = function(index, newNode) {
  if (!(newNode instanceof Node)) throw new MyException('input is not a Node');
  node = this.getNode(index, false);
  if (node === null) {
    this.addNode(newNode);
    return this;
  } else if (this.head == node) {
    newNode.next = node;
    node.prev = newNode;
    this.head = newNode;
  } else {
    node.prev.next = newNode;
    node.prev = newNode;
    newNode.next = node;
  }
  this.length += 1;
  return this;
};

LinkedList.prototype.deleteAtIndex = function(index) {
  node = this.getNode(index);
  if (index === 0) {
    if (this.length == 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = this.getNode(1);
    }
  }
  if (node.prev) node.prev.next = node.next;
  if (node.next) node.next.prev = node.prev;
  this.length -= 1;
  node.next = null;
  node.prev = null;
  node.value = null;
  return this;
};

module.exports = {
  Node: Node,
  LinkedList: LinkedList,
  MyException: MyException
};

File: python/linkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None


class LinkedList:
    def __init__(self):
        self.length = 0
        self.head = None
        self.tail = None

    def add_node(self, node):
        if not isinstance(node, Node):
            raise 'input value is not a Node'
        self.length += 1
        if not self.head:
            self.head = node
            self.tail = node
        else:
            node.prev = self.tail
            self.tail.next = node
            self.tail = node
        return self

    def remove_last(self):
        node = self.head
        if not node:
            raise 'Linked List is already empty'
        if not node.next:
            self.head = None
            self.tail = None
        else:
            while node.next:
                node = node.next
            node.prev.next = None
            self.tail = node.prev
        self.length -= 1
        return self

    def clean(self):
        def recursiveRm(self, node):
            if node.next:
                node.next.prev = None
            node.next = None
            node.value = None
            self.length -= 1
            if not node.prev:
                node.value = None
                return 0
            return recursiveRm(self, node.prev)
        node = self.tail
        if not node:
            raise 'Linked List is already empty'
        recursiveRm(self, node)
        self.head = None
        self.tail = None
        return self

    def get_node(self, index, notSkip=True):
        node = self.head
        if notSkip and not node:
            raise 'Linked List is already empty'
        if index > self.length or index < 0:
            raise 'invalid index input'
        for i in range(index):
            node = node.next
        return node

    def insert_at_index(self):
        pass

    def delete_at_index(self, index):
        node = self.get_node(index)
        if index == 0:
            if self.length == 1:
                self.tail = None
                self.head = None
            else:
                self.head = self.get_node(1)
        if node.prev:
            node.prev.next = node.next
        if node.next:
            node.next.prev = node.prev
        self.length -= 1
        node.next = None
        node.prev = None
        node.value = None
        return self
